drain3/drain.py

```diff
@@ -175,14 +185,17 @@ def get_seq_distance(seq1, seq2):
 
         return ret_val, param_count
 
-    def fast_match(self, cluster_list: list, tokens):
+    def fast_match(self, cluster_ids: list, tokens):
         match_cluster = None
 
         max_sim = -1
         max_param_count = -1
         max_cluster = None
 
-        for cluster in cluster_list:
+        for cluster_key in cluster_ids:
+            cluster = self.clusters.get(cluster_key)
```

Comment:
Yes, I did check performance implications. There is 15% impact on the perf mainly due to the clean-up of stale keys to clusters.

Since original Drain is exceptionally fast I think this is a good trade-off because this 15% gives us stable algorithm which can work with potentially unbounded los streams. Of course, this is just my opinion.

```
>pipenv run time -l python -m examples.drain_bigfile_demo
```

```
6.76 real         6.66 user         0.07 sys
    15998976  maximum resident set size
           0  average shared memory size
           0  average unshared data size
           0  average unshared stack size
        4505  page reclaims
           0  page faults
           0  swaps
           0  block input operations
           0  block output operations
           0  messages sent
           0  messages received
           0  signals received
           2  voluntary context switches
        1056  involuntary context switches
 57816017492  instructions retired
 26965436034  cycles elapsed
    13312000  peak memory footprint
```

```
7.77 real         7.65 user         0.08 sys
    16113664  maximum resident set size
           0  average shared memory size
           0  average unshared data size
           0  average unshared stack size
        4559  page reclaims
           0  page faults
           0  swaps
           0  block input operations
           0  block output operations
           0  messages sent
           0  messages received
           0  signals received
          38  voluntary context switches
        1937  involuntary context switches
 60512366022  instructions retired
 28471688577  cycles elapsed
    13426688  peak memory footprint
```

Reply:
I agree 15% is an acceptable trade-off